### DISCUSSION TITLE
In CellulOS what is a PD?

### DISCUSSION THREAD

**Discussion by sid-agrawal** [sid-agrawal/OSmosis#10] [General]
This has come up at many different times and again today while talking to @p-linh.

Is a CSpace a PD? Or is a {CSpace, CPU, VAS} a PD?

I will elaborate more on this later, but for now just wanted to open an issue, so that it is not lost.

**Comment by p-linh** [comment]
To document a related question that has also come up many times and came up again today: If PD1 = {CSpace1, CPU1, VAS1}, and we then slightly modified it to {CSpace1, CPU1, VAS2}, is it still PD1, or is this a separate PD2?

I thought about it a little bit, and the VAS currently seems to me to only be an interface for the PD to access its resources (and not something that defines it). E.g. if PD1 {CSpace1} has a cap to Resource1, and it switches from VAS1 to VAS2, it still has access to all of its resources, and RDEs. The only difference now is that it access those resources through a virtual addresses. 

I feel a similar argument could be made for the CPU object to not be a defining part of a PD. Although, it is still a bit head-scratching to me since we haven't yet decided if it just represents scheduler timeslices, or if it also includes an execution context.

**Reply by achreto** [reply]
> Is a CSpace a PD? Or is a {CSpace, CPU, VAS} a PD?

Maybe even neither.  

> we then slightly modified it to 

I'd say that this answer's the question: it would be a separate PD. We can't have a slightly modified PD; it's either the same or a different one. 

> and it switches from VAS1 to VAS2,

That would mean that the PD has both, pages from VAS1 and VAS2 as resources in the PD. The switch would do nothing in the abstract model, or change the permissions of the hold edges.
